### DISCUSSION TITLE
How can i use ilike operator for string type?

### DISCUSSION THREAD

**Discussion by divya-clari** [react-querybuilder/react-querybuilder#350] [Q&A] [answered]
The contains, beginswith, endswith operators are all internally transformed into like operator. I would like to do something similar to like operator which will be case insensitive. How can i achieve that ? any pointers will be helpful.

**Comment by jakeboone02** [comment, marked as answer]
I'm going to assume you're referring to the [`formatQuery`](https://react-querybuilder.js.org/docs/api/export) function.

Case sensitivity can be handled in different ways for each SQL implementation, so let me know which database product you're using (Oracle, MySQL, PostgreSQL, etc.) and I can help more.

The most generic way to make a query clause case-insensitive is to `UPPER()` or `LOWER()` both sides of the comparison. `formatQuery` doesn't actually provide a granular way to modify the left-hand side (the field/column name) in the SQL export, _however_, if you can assume that the data to be filtered is already uppercase or lowercase, then you only need to wrap the right-hand side in a function. You can provide a custom `valueProcessor` function to `formatQuery` which can modify the way a value is represented in the final string.

Let me know if you need some example code.

Edit: actual answer is https://github.com/react-querybuilder/react-querybuilder/discussions/350#discussioncomment-3329156

**Reply by divya-clari** [reply]
Thanks for the brief reply. The database is PostgreSQL. In my case we cannot assume the data is already filtered. So i don't think wrapping the right-hand side alone will help. I have to wrap both the field name and the value if i have to use like operator. for example `where upper(name) like upper(%testName) `. Any pointers for wrapping the field name will be helpful. if not any guidance on how i can use ilike operator will help too! A sample code will definitely help. Thank's once again

**Reply by jakeboone02** [reply]
I'm looking into this. I don't have a good answer right now but maybe will in a day or two.

**Reply by divya-clari** [reply]
ok thank you very much. We are using the query builder extensively and the support for this use case is critical to us. Appreciate your prompt response and all your efforts in maintaining this component.

**Reply by jakeboone02** [reply]
Ok this isn't the most elegant solution, but I think it will meet your immediate requirement.

First, I augmented the default operator list with alternate, case-insensitive operators:

```ts
const operators: OptionGroup<NameLabelPair>[] = [
  { label: "Case Sensitive", options: defaultOperators },
  {
    label: "Case Insensitive",
    options: [
      { name: "contains-ci", label: "contains (ci)" },
      { name: "beginsWith-ci", label: "begins with (ci)" },
      { name: "endsWith-ci", label: "ends with (ci)" },
      { name: "doesNotContain-ci", label: "does not contain (ci)" },
      { name: "doesNotBeginWith-ci", label: "does not begin with (ci)" },
      { name: "doesNotEndWith-ci", label: "does not end with (ci)" }
    ]
  }
];
```

Next, I set up a `valueProcessor` that would send each rule through the `defaultValueProcessor` as if the operator was its standard, case-sensitive counterpart. For example, `"contains-ci"` would get translated to `"contains"`. This inserts the `%` where necessary.

```ts
const valueProcessor: ValueProcessorByRule = (rule, opts) =>
  defaultValueProcessorByRule(
    { ...rule, operator: rule.operator.replace("-ci", "") },
    opts
  );

const fqOptions: FormatQueryOptions = { format: "sql", valueProcessor };
```

Next, I replaced all the `"*-ci"` operators in the string returned by `formatQuery` with `"ILIKE"`.

```ts
formatQuery(query, fqOptions).replace(/\b[a-zA-Z]+-ci\b/gi, "ILIKE")
```

That turns a query like this:

```json
{
  "combinator": "and",
  "rules": [
    { "field": "firstName", "operator": "contains-ci", "value": "abc" },
    { "field": "lastName", "operator": "doesNotBeginWith-ci", "value": "xyz" }
  ]
}
```

...into this:

```
(firstName ILIKE '%abc%' and lastName ILIKE 'xyz%')
```

I would eventually like to provide a configurable way to manipulate the left-hand side and the operator, but this may be the best we can do for now.

**Reply by jakeboone02** [reply]
Oops! Forgot to put the link to the codesandbox: https://codesandbox.io/s/lucid-nightingale-snom76?file=/src/App.tsx

**Reply by divya-clari** [reply]
Thanks @jakeboone02 for the quick workaround. This is really helpful. One follow up question. I also use the parseSql method to read from SQL. So how can i make it support the Ilike operator. 
```
 <QueryBuilder 
            enableDragAndDrop={true}
            fields= {localFields}
            defaultQuery = {model.read != '()' ? parseSQL(model.read) : query}
            onQueryChange={q => modelUpdate({
              query: formatQuery(q, {format:'sql', 
                                     fields: localFields ,
                                     valueProcessor: valueProcessor}).replace(/\b[a-zA-Z]+-ci\b/gi, "ILIKE")})}
            combinators={model.combinators}
            resetOnFieldChange = {true}
            controlElements = {{ 
              valueEditor: customValueEditor}}
 />
```

**Reply by jakeboone02** [reply]
That one's not actually extensible since it only really covers ANSI SQL operators. `ILIKE` is specific to Postgres, and on top of that there are no corresponding operators in the `defaultOperators` list that I could map it to.

**Reply by jakeboone02** [reply]
@divya-clari if https://github.com/react-querybuilder/react-querybuilder/discussions/350#discussioncomment-3329156 helped you out enough, want to mark this thread as answered?

**Reply by divya-clari** [reply]
@jakeboone02 Thanks for the reminder here! Yes will do

**Comment by divya-clari** [comment]
@jakeboone02 Thanks for the reply. I need another clarification on the same lines. Hope its ok to ask here. I have another operator defined similar to Is Null with no value. `{name: "<> \'\'", label: 'Is Present'},` Which basically checks if the column is not blank. So does this mean parsesql cannot support this use case as well.

**Reply by jakeboone02** [reply]
`parseSQL` doesn't handle any custom operators, only the standard ones that can map to the `defaultOperators` list. It _can_ pick up on "IS NOT NULL" and map it to the `"notNull"` operator. If it saw `f1 <> ''`, I believe that would just map to this:

```json
{ "field": "f1", "operator": "!=", "value": "" }
```
